src/extension.ts

```diff
@@ -48,45 +48,71 @@ async function fetchPackageInfo(requirement: PackageRequirement): Promise<Packag
         return [response.status, info]
     } catch (e) {
         return [null, null]
     }
 }
 
-function presentPackageInfo(info: PackageInfo): string {
+function presentPackageInfo(info: PackageInfo): string[] {
     const summarySubPart: string = info.summary ? ` – ${linkify(info.summary, info.home_page)}` : ''
     const headPart: string = `**${linkify(info.name, info.package_url)}${summarySubPart}**`
     const emailSubpart: string = info.author_email ? ` (${info.author_email})` : ''
     const authorSubpart: string = info.author && info.author_email ? `By ${info.author}${emailSubpart}.` : ''
     const licenseSubpart: string = info.license ? ` ${info.license.replace(/ licen[cs]e/gi, '')} licensed.` : ''
     const versionPart: string = `Latest version: ${linkify(info.version, info.release_url)}.`
-    const infoPresentation: string = [
-        headPart, authorSubpart + licenseSubpart, versionPart
-    ].filter(Boolean).join('\n\n')
+    const versionPartLens: string = `Latest version: ${info.version}`
+    const infoPresentation: Array<string> = [
+        headPart, authorSubpart + licenseSubpart, versionPart, versionPartLens
+    ].filter(Boolean)
+
     return infoPresentation
 }
 
 async function provideHover(document: vscode.TextDocument, position: vscode.Position) {
     const requirement: PackageRequirement | null = extractPackageRequirement(document.lineAt(position.line))
     if (requirement === null) return new vscode.Hover('')
-    let infoPresentation: string | undefined = infoPresentationCache.get(requirement.id)
-    if (infoPresentation === undefined) {
-        const [status, info]: PackageInfoRequest = await fetchPackageInfo(requirement)
-        if (status === 200) infoPresentation = presentPackageInfo(info!)
-        else if (status === 404) infoPresentation = `{id_raw} is not available in PyPI`
-        if (infoPresentation) infoPresentationCache.set(requirement.id, infoPresentation)
-        else infoPresentation = linkify(
-            `could not fetch ${requirement.id_raw} information from PyPI`,
-            `https://pypi.org/project/${requirement.id_raw}/`
-        )
-    }
-    infoPresentation = infoPresentation.replace('{id_raw}', requirement.id_raw)
-    return new vscode.Hover(infoPresentation)
+    let infoPresentation: Array<string> | undefined = infoPresentationCache.get(requirement.id)
+    if (infoPresentation === undefined) return new vscode.Hover('')
+    return new vscode.Hover(infoPresentation.slice(0,-1).join("\n\n").replace('{id_raw}', requirement.id_raw))
 }
 
+class MyCodeLensProvider implements vscode.CodeLensProvider {
+    async provideCodeLenses(document: vscode.TextDocument): Promise<vscode.CodeLens[]> {
+
+      const lineCount: number = document.lineCount;
+
+      const requirements: Array<PackageRequirement|null> = 
+      [...Array(lineCount).keys()]
+      .map(lineNumber => extractPackageRequirement(document.lineAt(lineNumber)))
```

Comment:
Critical: This line number-based matching isn't viable, because it's not at all guaranteed that every single line corresponds to a single requirement. Just look at what happens in this file which has a few comments in between requirements:

![Zrzut ekranu 2022-03-1 o 22 49 05](https://user-images.githubusercontent.com/4550621/156254643-65b4f182-31f8-47ab-b1d1-74f29997a110.png)

Reply:
![image](https://user-images.githubusercontent.com/30223404/156429498-32599e91-f94a-4b85-8c01-9b52e0619ef4.png)

Fixed. Now able to handle invalid lines.

Reply:
In Version Lens for `package.json`, requests were only made when `package.json` is formatted correctly, with version numbers for every package. However, for python `requirements.txt`, version numbers aren't mandatory. Currently, pypi lookup attempts are made every time a user types on a line, which can be wasteful.

Should we add a delay between search attempts?

Reply:
Yeah, that'd make sense. What we can do in my view is maintaining a map `Record<number, NodeJS.Time>` and make it so that once initial results for a line are in, `fetchPackageInfo` is debounced by 500 ms using that map. Do you want to take this up?

Reply:
Sure, I'll try it out.